{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GWAS summery stat data preprocess\n",
    "1.need to convert Markername to rsID by combining Chr and basepair coordinate in _lifted bim file and and replace Markername by matching\n",
    "2.calculate z-score using P-value and effect size using munge_sumstat in ldsc github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MarkerName      Allele1 Allele2 Freq1   FreqSE  MinFreq MaxFreq Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal freqSpan\n",
      "chr11:88249377  t       c       0.9905  0.0026  0.9882  0.9971  0.0238  0.0594  0.6889  -+?+++--+???-+++?       20      13.753  11      0.247   0.00890000000000002\n",
      "chr1:60320992   a       g       0.9375  0.0049  0.9221  0.9473  0.0185  0.021   0.3801  +-?--+?++??-+---+       20.1    15.022  12      0.2402  0.0252\n",
      "chr8:135908647  a       g       0.2112  0.007   0.1703  0.2316  -0.0054 0.0123  0.6598  --?+++?--??-----+       27.2    16.492  12      0.1697  0.0613\n",
      "chr16:77148858  a       g       0.9963  5e-04   0.9961  0.9984  -0.0228 0.1356  0.8663  -??-+??+-???----?       0       4.331   8       0.8261  0.00229999999999997\n",
      "chr11:97895884  c       g       0.0587  0.0035  0.0361  0.0685  -0.0219 0.0217  0.3148  +-?--+?--??+---++       0.2     12.029  12      0.4433  0.0324\n",
      "chr2:80117945   a       g       0.9979  7e-04   0.9968  0.9988  -0.4366 0.4506  0.3326  ?????+?--???---+?       0       5.236   6       0.5139  0.002\n",
      "chr18:814730    c       g       0.8893  0.0075  0.8639  0.9086  -0.0022 0.0176  0.9025  --?---?+-??+++--+       0       4.652   12      0.9686  0.0447\n",
      "chr19:32978300  a       g       0.0043  9e-04   0.0034  0.0075  0.0398  0.0991  0.6879  +-?--??--???+--+?       0       4.429   9       0.881   0.0041\n",
      "chr2:166277422  a       g       0.0023  0.0022  0.0011  0.0076  -0.1177 0.308   0.7023  ???+---?+????+???       21.5    6.372   5       0.2717  0.0065\n",
      "\u001b[K:\u001b[K20407145 t       c       0.9112  0.0071  0.8824  0.9202  -0.0278 0.0179  \u001b[7m/data/CARD/PD/summary_stats/META5_all.gz\u001b[m\u001b[K\u0007"
     ]
    }
   ],
   "source": [
    "#read meta5 file -- this is our GWAS summery stat data\n",
    "! zless /data/CARD/PD/summary_stats/META5_all.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy meta 5 file into data folder - run only once for download\n",
    "! cp -R /data/CARD/PD/summary_stats/META5_all.gz /data/songy4/TWAS/data/\n",
    "\n",
    "#unzip META5_all.gz - run only once for download\n",
    "! gzip -d META5_all.gz\n",
    "\n",
    "#make sumstat.csv which contains MarkerName, Allele1, Allele2, Effect, StdErr, P-value\n",
    "! awk '{print $1, $2, $3, $8, $9, $10}' data_folder/META5_all > data_folder/sumstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import library\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#set paths\n",
    "basedir = '/data/songy4/twas'\n",
    "datadir = f'{basedir}/data_folder'\n",
    "fusiondir = f'{basedir}/fusion_twas'\n",
    "geno_path = f'{datadir}/qc_genotypes_twas'\n",
    "gene_list_path = f'{datadir}/gene_list.txt'\n",
    "pheno_path = f'{datadir}/expression_matrix_final.txt'\n",
    "coord_path = f'{datadir}/twas_coordinate.txt'\n",
    "covar_path = f'{datadir}/covariates_case.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>FreqSE</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>freqSpan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.9882</td>\n",
       "      <td>0.9971</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "      <td>-+?+++--+???-+++?</td>\n",
       "      <td>20.0</td>\n",
       "      <td>13.753</td>\n",
       "      <td>11</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.0089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.9221</td>\n",
       "      <td>0.9473</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>+-?--+?++??-+---+</td>\n",
       "      <td>20.1</td>\n",
       "      <td>15.022</td>\n",
       "      <td>12</td>\n",
       "      <td>0.2402</td>\n",
       "      <td>0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>0.0070</td>\n",
       "      <td>0.1703</td>\n",
       "      <td>0.2316</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>--?+++?--??-----+</td>\n",
       "      <td>27.2</td>\n",
       "      <td>16.492</td>\n",
       "      <td>12</td>\n",
       "      <td>0.1697</td>\n",
       "      <td>0.0613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9963</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "      <td>-??-+??+-???----?</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.331</td>\n",
       "      <td>8</td>\n",
       "      <td>0.8261</td>\n",
       "      <td>0.0023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0685</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.3148</td>\n",
       "      <td>+-?--+?--??+---++</td>\n",
       "      <td>0.2</td>\n",
       "      <td>12.029</td>\n",
       "      <td>12</td>\n",
       "      <td>0.4433</td>\n",
       "      <td>0.0324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName Allele1 Allele2   Freq1  FreqSE  MinFreq  MaxFreq  Effect  \\\n",
       "0  chr11:88249377       t       c  0.9905  0.0026   0.9882   0.9971  0.0238   \n",
       "1   chr1:60320992       a       g  0.9375  0.0049   0.9221   0.9473  0.0185   \n",
       "2  chr8:135908647       a       g  0.2112  0.0070   0.1703   0.2316 -0.0054   \n",
       "3  chr16:77148858       a       g  0.9963  0.0005   0.9961   0.9984 -0.0228   \n",
       "4  chr11:97895884       c       g  0.0587  0.0035   0.0361   0.0685 -0.0219   \n",
       "\n",
       "   StdErr  P-value          Direction  HetISq  HetChiSq  HetDf  HetPVal  \\\n",
       "0  0.0594   0.6889  -+?+++--+???-+++?    20.0    13.753     11   0.2470   \n",
       "1  0.0210   0.3801  +-?--+?++??-+---+    20.1    15.022     12   0.2402   \n",
       "2  0.0123   0.6598  --?+++?--??-----+    27.2    16.492     12   0.1697   \n",
       "3  0.1356   0.8663  -??-+??+-???----?     0.0     4.331      8   0.8261   \n",
       "4  0.0217   0.3148  +-?--+?--??+---++     0.2    12.029     12   0.4433   \n",
       "\n",
       "   freqSpan  \n",
       "0    0.0089  \n",
       "1    0.0252  \n",
       "2    0.0613  \n",
       "3    0.0023  \n",
       "4    0.0324  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statsum = pd.read_csv(r\"./data_folder/META5_all\", sep='\\t')\n",
    "statsum.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4973527\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>FreqSE</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>freqSpan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.9882</td>\n",
       "      <td>0.9971</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.688900</td>\n",
       "      <td>-+?+++--+???-+++?</td>\n",
       "      <td>20.0</td>\n",
       "      <td>13.753</td>\n",
       "      <td>11</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.0089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.9221</td>\n",
       "      <td>0.9473</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>+-?--+?++??-+---+</td>\n",
       "      <td>20.1</td>\n",
       "      <td>15.022</td>\n",
       "      <td>12</td>\n",
       "      <td>0.2402</td>\n",
       "      <td>0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9963</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.866300</td>\n",
       "      <td>-??-+??+-???----?</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.331</td>\n",
       "      <td>8</td>\n",
       "      <td>0.8261</td>\n",
       "      <td>0.0023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2:80117945</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9979</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.9968</td>\n",
       "      <td>0.9988</td>\n",
       "      <td>-0.4366</td>\n",
       "      <td>0.4506</td>\n",
       "      <td>0.332600</td>\n",
       "      <td>?????+?--???---+?</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.236</td>\n",
       "      <td>6</td>\n",
       "      <td>0.5139</td>\n",
       "      <td>0.0020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18:814730</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.8893</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.8639</td>\n",
       "      <td>0.9086</td>\n",
       "      <td>-0.0022</td>\n",
       "      <td>0.0176</td>\n",
       "      <td>0.902500</td>\n",
       "      <td>--?---?+-??+++--+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.652</td>\n",
       "      <td>12</td>\n",
       "      <td>0.9686</td>\n",
       "      <td>0.0447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11477537</th>\n",
       "      <td>chr2:38455448</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9929</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.9867</td>\n",
       "      <td>0.9969</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.0667</td>\n",
       "      <td>0.967600</td>\n",
       "      <td>-+?-++?-+?+?-++-?</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.288</td>\n",
       "      <td>11</td>\n",
       "      <td>0.8534</td>\n",
       "      <td>0.0102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11477538</th>\n",
       "      <td>chr7:65912486</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.9977</td>\n",
       "      <td>0.9988</td>\n",
       "      <td>0.8050</td>\n",
       "      <td>0.6315</td>\n",
       "      <td>0.202400</td>\n",
       "      <td>????+??+-????++??</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.030</td>\n",
       "      <td>4</td>\n",
       "      <td>0.7303</td>\n",
       "      <td>0.0011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11477540</th>\n",
       "      <td>chr18:50482478</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9722</td>\n",
       "      <td>0.0023</td>\n",
       "      <td>0.9661</td>\n",
       "      <td>0.9807</td>\n",
       "      <td>-0.0203</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>0.525300</td>\n",
       "      <td>--?-+-+-+??+++-+-</td>\n",
       "      <td>12.8</td>\n",
       "      <td>14.913</td>\n",
       "      <td>13</td>\n",
       "      <td>0.3128</td>\n",
       "      <td>0.0146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11477541</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9452</td>\n",
       "      <td>0.0034</td>\n",
       "      <td>0.9326</td>\n",
       "      <td>0.9763</td>\n",
       "      <td>-0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>--+++++-+++--+-++</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.081</td>\n",
       "      <td>16</td>\n",
       "      <td>0.9101</td>\n",
       "      <td>0.0437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11477542</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.6499</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>0.6378</td>\n",
       "      <td>0.6924</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>+++++-++++---+-+-</td>\n",
       "      <td>25.1</td>\n",
       "      <td>21.354</td>\n",
       "      <td>16</td>\n",
       "      <td>0.1653</td>\n",
       "      <td>0.0546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4973526 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName Allele1 Allele2   Freq1  FreqSE  MinFreq  MaxFreq  \\\n",
       "0         chr11:88249377       t       c  0.9905  0.0026   0.9882   0.9971   \n",
       "1          chr1:60320992       a       g  0.9375  0.0049   0.9221   0.9473   \n",
       "3         chr16:77148858       a       g  0.9963  0.0005   0.9961   0.9984   \n",
       "5          chr2:80117945       a       g  0.9979  0.0007   0.9968   0.9988   \n",
       "6           chr18:814730       c       g  0.8893  0.0075   0.8639   0.9086   \n",
       "...                  ...     ...     ...     ...     ...      ...      ...   \n",
       "11477537   chr2:38455448       t       c  0.9929  0.0010   0.9867   0.9969   \n",
       "11477538   chr7:65912486       t       g  0.9981  0.0005   0.9977   0.9988   \n",
       "11477540  chr18:50482478       c       g  0.9722  0.0023   0.9661   0.9807   \n",
       "11477541  chr11:47906648       t       c  0.9452  0.0034   0.9326   0.9763   \n",
       "11477542   chr3:47855594       t       c  0.6499  0.0093   0.6378   0.6924   \n",
       "\n",
       "          Effect  StdErr   P-value          Direction  HetISq  HetChiSq  \\\n",
       "0         0.0238  0.0594  0.688900  -+?+++--+???-+++?    20.0    13.753   \n",
       "1         0.0185  0.0210  0.380100  +-?--+?++??-+---+    20.1    15.022   \n",
       "3        -0.0228  0.1356  0.866300  -??-+??+-???----?     0.0     4.331   \n",
       "5        -0.4366  0.4506  0.332600  ?????+?--???---+?     0.0     5.236   \n",
       "6        -0.0022  0.0176  0.902500  --?---?+-??+++--+     0.0     4.652   \n",
       "...          ...     ...       ...                ...     ...       ...   \n",
       "11477537  0.0027  0.0667  0.967600  -+?-++?-+?+?-++-?     0.0     6.288   \n",
       "11477538  0.8050  0.6315  0.202400  ????+??+-????++??     0.0     2.030   \n",
       "11477540 -0.0203  0.0319  0.525300  --?-+-+-+??+++-+-    12.8    14.913   \n",
       "11477541 -0.0062  0.0207  0.765700  --+++++-+++--+-++     0.0     9.081   \n",
       "11477542  0.0348  0.0105  0.000944  +++++-++++---+-+-    25.1    21.354   \n",
       "\n",
       "          HetDf  HetPVal  freqSpan  \n",
       "0            11   0.2470    0.0089  \n",
       "1            12   0.2402    0.0252  \n",
       "3             8   0.8261    0.0023  \n",
       "5             6   0.5139    0.0020  \n",
       "6            12   0.9686    0.0447  \n",
       "...         ...      ...       ...  \n",
       "11477537     11   0.8534    0.0102  \n",
       "11477538      4   0.7303    0.0011  \n",
       "11477540     13   0.3128    0.0146  \n",
       "11477541     16   0.9101    0.0437  \n",
       "11477542     16   0.1653    0.0546  \n",
       "\n",
       "[4973526 rows x 16 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#count if Freq1 is bigger than 0.5 --> this is to check if minor allele is A1\n",
    "!cat {datadir}/META5_all | awk -F\"\\t\" '$4>0.5' | wc -l\n",
    "\n",
    "#check again how many Freq1 \n",
    "statsum_flip = statsum[statsum.Freq1 > 0.5]\n",
    "statsum_flip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique MarkerName in meta data: 7784413\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.314800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18:814730</td>\n",
       "      <td>g</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0022</td>\n",
       "      <td>0.0176</td>\n",
       "      <td>0.902500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784408</th>\n",
       "      <td>chr18:50482478</td>\n",
       "      <td>g</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0203</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>0.525300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784409</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784410</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784411</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784412</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7784413 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName Allele1 Allele2  Effect  StdErr   P-value\n",
       "0          chr1:60320992       g       a -0.0185  0.0210  0.380100\n",
       "1         chr8:135908647       a       g -0.0054  0.0123  0.659800\n",
       "2         chr11:97895884       c       g -0.0219  0.0217  0.314800\n",
       "3           chr18:814730       g       c  0.0022  0.0176  0.902500\n",
       "4        chr10:120407145       c       t  0.0278  0.0179  0.120300\n",
       "...                  ...     ...     ...     ...     ...       ...\n",
       "7784408   chr18:50482478       g       c  0.0203  0.0319  0.525300\n",
       "7784409   chr11:47906648       c       t  0.0062  0.0207  0.765700\n",
       "7784410    chr3:47855594       c       t -0.0348  0.0105  0.000944\n",
       "7784411   chr3:194372962       a       g  0.0268  0.0500  0.591900\n",
       "7784412   chr19:34986930       a       c  0.0209  0.0154  0.174900\n",
       "\n",
       "[7784413 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#flip Allele1 and Allele2 if Freq1 is bigger than .5 and change Freq1 to negative\n",
    "stat_sum = statsum.copy()\n",
    "cond = (stat_sum['Freq1'] > 0.5)\n",
    "stat_sum.loc[cond, ['Allele1', 'Allele2']] = stat_sum.loc[cond, ['Allele2', 'Allele1']].values\n",
    "#stat_sum_flip.Allele1, stat_sum_flip.Allele2 = np.where(stat_sum_flip.Freq1 > 0.5, [stat_sum_flip.Allele2, stat_sum_flip.Allele1], [stat_sum_flip.Allele1, stat_sum_flip.Allele2])\n",
    "\n",
    "#change Freq1\n",
    "stat_sum.loc[cond, ['Freq1']] = 1-stat_sum['Freq1']\n",
    "\n",
    "#change Effect value\n",
    "stat_sum.loc[cond, 'Effect'] *= -1\n",
    "##no need to change  StdErr, P-value\n",
    "\n",
    "#drop two rows Freq1 is smaller than 0.01\n",
    "stat_sum.drop(stat_sum[stat_sum.Freq1 < 0.01].index, inplace=True)\n",
    "stat_sum = stat_sum.reset_index(drop=True)\n",
    "#stat_sum = stat_sum.drop(['index'], axis=1)\n",
    "\n",
    "#drop unwanted columns\n",
    "stat_sum.drop(stat_sum.iloc[:,10:16], axis=1, inplace=True)\n",
    "stat_sum.drop(stat_sum.iloc[:,3:7], axis=1, inplace=True)\n",
    "\n",
    "#check the unique number of MarkerName in meta\n",
    "print(\"Unique MarkerName in meta data:\", stat_sum['MarkerName'].nunique())\n",
    "\n",
    "stat_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save stat_sum \n",
    "stat_sum.to_csv(r'./data_folder/sumstat.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shell_do(command, log=False, return_log=False):\n",
    "    print(f'Executing: {(\" \").join(command.split())}', file=sys.stderr)\n",
    "\n",
    "    res=subprocess.run(command.split(), stdout=subprocess.PIPE)\n",
    "\n",
    "    if log:\n",
    "        print(res.stdout.decode('utf-8'))\n",
    "    if return_log:\n",
    "        return(res.stdout.decode('utf-8'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of bim data: (23060234, 7)\n",
      "number of unique MarkerName in bim data: 23060188\n",
      "number of unique rsid in bim data: 23060234\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs145427775</td>\n",
       "      <td>0</td>\n",
       "      <td>10291</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs55998931</td>\n",
       "      <td>0</td>\n",
       "      <td>10492</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199896944</td>\n",
       "      <td>0</td>\n",
       "      <td>13504</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:13504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>0</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0</td>\n",
       "      <td>14948</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr         rsid  kb    pos a1 a2  MarkerName\n",
       "0    1  rs145427775   0  10291  T  C  chr1:10291\n",
       "1    1   rs55998931   0  10492  T  C  chr1:10492\n",
       "2    1  rs199896944   0  13504  A  G  chr1:13504\n",
       "3    1  rs199856693   0  14933  A  G  chr1:14933\n",
       "4    1  rs201855936   0  14948  A  G  chr1:14948"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open bim file \n",
    "bim_df = pd.read_csv(f\"{geno_path}_hg19_lifted.bim\", sep='\\t', header=None,  names=('chr', 'rsid', 'kb', 'pos', 'a1', 'a2'))\n",
    "\n",
    "#combine chr and pos columns\n",
    "bim_df['MarkerName'] = str('chr') + bim_df.chr.astype(str) + ':' + bim_df.pos.astype(str)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"shape of bim data:\", bim_df.shape)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"number of unique MarkerName in bim data:\", bim_df['MarkerName'].nunique())\n",
    "print(\"number of unique rsid in bim data:\", bim_df['rsid'].nunique())\n",
    "bim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of unique MarkerName in duplicated bim_df:  46\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>978590</th>\n",
       "      <td>1</td>\n",
       "      <td>rs368379639</td>\n",
       "      <td>0</td>\n",
       "      <td>148609803</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:148609803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978719</th>\n",
       "      <td>1</td>\n",
       "      <td>rs372742293</td>\n",
       "      <td>0</td>\n",
       "      <td>148561459</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:148561459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>979275</th>\n",
       "      <td>1</td>\n",
       "      <td>rs377761119</td>\n",
       "      <td>0</td>\n",
       "      <td>148609803</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:148609803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991861</th>\n",
       "      <td>1</td>\n",
       "      <td>rs111636993</td>\n",
       "      <td>0</td>\n",
       "      <td>148561459</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:148561459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5405594</th>\n",
       "      <td>3</td>\n",
       "      <td>rs368733146</td>\n",
       "      <td>0</td>\n",
       "      <td>195218551</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr3:195218551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22766969</th>\n",
       "      <td>22</td>\n",
       "      <td>rs188638512</td>\n",
       "      <td>0</td>\n",
       "      <td>19669438</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22:19669438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22767437</th>\n",
       "      <td>22</td>\n",
       "      <td>rs374752842</td>\n",
       "      <td>0</td>\n",
       "      <td>19719099</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22:19719099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22767608</th>\n",
       "      <td>22</td>\n",
       "      <td>rs41295659</td>\n",
       "      <td>0</td>\n",
       "      <td>19738319</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22:19738319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22767625</th>\n",
       "      <td>22</td>\n",
       "      <td>rs41295687</td>\n",
       "      <td>0</td>\n",
       "      <td>19739438</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr22:19739438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22768084</th>\n",
       "      <td>22</td>\n",
       "      <td>rs200296503</td>\n",
       "      <td>0</td>\n",
       "      <td>19786130</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTCCTGCCCAG</td>\n",
       "      <td>chr22:19786130</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr         rsid  kb        pos a1            a2      MarkerName\n",
       "978590      1  rs368379639   0  148609803  A             G  chr1:148609803\n",
       "978719      1  rs372742293   0  148561459  T             C  chr1:148561459\n",
       "979275      1  rs377761119   0  148609803  A             G  chr1:148609803\n",
       "991861      1  rs111636993   0  148561459  T             C  chr1:148561459\n",
       "5405594     3  rs368733146   0  195218551  T             G  chr3:195218551\n",
       "...       ...          ...  ..        ... ..           ...             ...\n",
       "22766969   22  rs188638512   0   19669438  T             C  chr22:19669438\n",
       "22767437   22  rs374752842   0   19719099  T             C  chr22:19719099\n",
       "22767608   22   rs41295659   0   19738319  A             C  chr22:19738319\n",
       "22767625   22   rs41295687   0   19739438  A             G  chr22:19739438\n",
       "22768084   22  rs200296503   0   19786130  C  CCTCCTGCCCAG  chr22:19786130\n",
       "\n",
       "[92 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print duplicated rows\n",
    "bim_df_dup = bim_df[bim_df['MarkerName'].duplicated(keep=False)]\n",
    "print('number of unique MarkerName in duplicated bim_df: ', bim_df_dup['MarkerName'].nunique())\n",
    "bim_df_dup\n",
    "## there are 46 duplicated MarkerName with 2 different rsid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique allele 1 in duplicated bim_df:  ['A' 'T' 'C' '*' 'G' 'AAGAC']\n",
      "Unique allele 2 in duplicated bim_df:  ['G' 'C' 'T' 'ACC' 'TG' 'A' 'AC' 'ACAG' 'CCTCCTGCCCAG']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3617568</th>\n",
       "      <td>chr22:19503090</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0449</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.48050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3861458</th>\n",
       "      <td>chr22:19172321</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.0183</td>\n",
       "      <td>0.52360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3939528</th>\n",
       "      <td>chr22:19497230</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.07624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4745961</th>\n",
       "      <td>chr22:19390804</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0161</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>0.32640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6030528</th>\n",
       "      <td>chr22:19643104</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0006</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.96150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             MarkerName Allele1 Allele2  Effect  StdErr  P-value\n",
       "3617568  chr22:19503090       c       t -0.0449  0.0637  0.48050\n",
       "3861458  chr22:19172321       t       g  0.0117  0.0183  0.52360\n",
       "3939528  chr22:19497230       c       t -0.0466  0.0263  0.07624\n",
       "4745961  chr22:19390804       c       t -0.0161  0.0164  0.32640\n",
       "6030528  chr22:19643104       c       t -0.0006  0.0134  0.96150"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Unique allele 1 in duplicated bim_df: ', bim_df_dup['a1'].unique())\n",
    "print('Unique allele 2 in duplicated bim_df: ', bim_df_dup['a2'].unique())\n",
    "#duplicated marker name\n",
    "dup_marker = bim_df_dup['MarkerName'].unique().tolist()\n",
    "\n",
    "#print gwas summery stat that has duplicated MarkerName\n",
    "stat_sum_dup = stat_sum[stat_sum['MarkerName'].isin(dup_marker)]\n",
    "stat_sum_dup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22740222</th>\n",
       "      <td>22</td>\n",
       "      <td>rs199687134</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19172321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22762532</th>\n",
       "      <td>22</td>\n",
       "      <td>rs1780641</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr22:19172321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22740918</th>\n",
       "      <td>22</td>\n",
       "      <td>rs141585725</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22:19390804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22764321</th>\n",
       "      <td>22</td>\n",
       "      <td>rs11089274</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19390804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22741203</th>\n",
       "      <td>22</td>\n",
       "      <td>rs2521701</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr22:19497230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22765096</th>\n",
       "      <td>22</td>\n",
       "      <td>rs117551893</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19497230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22741234</th>\n",
       "      <td>22</td>\n",
       "      <td>rs370330113</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19503090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22765143</th>\n",
       "      <td>22</td>\n",
       "      <td>rs138605059</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19503090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22741527</th>\n",
       "      <td>22</td>\n",
       "      <td>rs145711952</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr22:19643104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22766688</th>\n",
       "      <td>22</td>\n",
       "      <td>rs9618649</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:19643104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr         rsid  kb       pos a1 a2      MarkerName\n",
       "22740222   22  rs199687134   0  19172321  C  T  chr22:19172321\n",
       "22762532   22    rs1780641   0  19172321  T  G  chr22:19172321\n",
       "22740918   22  rs141585725   0  19390804  T  C  chr22:19390804\n",
       "22764321   22   rs11089274   0  19390804  C  T  chr22:19390804\n",
       "22741203   22    rs2521701   0  19497230  T  C  chr22:19497230\n",
       "22765096   22  rs117551893   0  19497230  C  T  chr22:19497230\n",
       "22741234   22  rs370330113   0  19503090  C  T  chr22:19503090\n",
       "22765143   22  rs138605059   0  19503090  C  T  chr22:19503090\n",
       "22741527   22  rs145711952   0  19643104  A  G  chr22:19643104\n",
       "22766688   22    rs9618649   0  19643104  C  T  chr22:19643104"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print bim_df_dup that has duplicated MarkerName to see rsid\n",
    "double_marker  = stat_sum_dup['MarkerName'].tolist()\n",
    "bim_df_dup_exist = bim_df_dup[bim_df_dup['MarkerName'].isin(double_marker)]\n",
    "\n",
    "#sort the dataframe by MarkerName\n",
    "bim_df_dup_exist = bim_df_dup_exist.sort_values(by = 'MarkerName')\n",
    "\n",
    "bim_df_dup_exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>0</td>\n",
       "      <td>60320992</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>0</td>\n",
       "      <td>135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>0</td>\n",
       "      <td>120407145</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.822100</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>0</td>\n",
       "      <td>29372878</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.214100</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>0</td>\n",
       "      <td>164053059</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478463</th>\n",
       "      <td>chr16:84525085</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.211700</td>\n",
       "      <td>16</td>\n",
       "      <td>rs2927754</td>\n",
       "      <td>0</td>\n",
       "      <td>84525085</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478464</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>11</td>\n",
       "      <td>rs148423893</td>\n",
       "      <td>0</td>\n",
       "      <td>47906648</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478465</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>3</td>\n",
       "      <td>rs62263573</td>\n",
       "      <td>0</td>\n",
       "      <td>47855594</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478466</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "      <td>3</td>\n",
       "      <td>rs149847890</td>\n",
       "      <td>0</td>\n",
       "      <td>194372962</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478467</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "      <td>19</td>\n",
       "      <td>rs13353264</td>\n",
       "      <td>0</td>\n",
       "      <td>34986930</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6478468 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName Allele1 Allele2  Effect  StdErr   P-value  chr  \\\n",
       "0          chr1:60320992       g       a -0.0185  0.0210  0.380100    1   \n",
       "1         chr8:135908647       a       g -0.0054  0.0123  0.659800    8   \n",
       "2        chr10:120407145       c       t  0.0278  0.0179  0.120300   10   \n",
       "3         chr11:29372878       t       c  0.0049  0.0218  0.822100   11   \n",
       "4         chr3:164053059       a       g  0.0607  0.0488  0.214100    3   \n",
       "...                  ...     ...     ...     ...     ...       ...  ...   \n",
       "6478463   chr16:84525085       t       c -0.0387  0.0310  0.211700   16   \n",
       "6478464   chr11:47906648       c       t  0.0062  0.0207  0.765700   11   \n",
       "6478465    chr3:47855594       c       t -0.0348  0.0105  0.000944    3   \n",
       "6478466   chr3:194372962       a       g  0.0268  0.0500  0.591900    3   \n",
       "6478467   chr19:34986930       a       c  0.0209  0.0154  0.174900   19   \n",
       "\n",
       "                rsid  kb        pos a1 a2  \n",
       "0        rs116406626   0   60320992  G  A  \n",
       "1         rs11992603   0  135908647  A  G  \n",
       "2          rs4336949   0  120407145  C  T  \n",
       "3         rs72638174   0   29372878  T  C  \n",
       "4        rs115079612   0  164053059  A  G  \n",
       "...              ...  ..        ... .. ..  \n",
       "6478463    rs2927754   0   84525085  T  C  \n",
       "6478464  rs148423893   0   47906648  C  T  \n",
       "6478465   rs62263573   0   47855594  C  T  \n",
       "6478466  rs149847890   0  194372962  A  G  \n",
       "6478467   rs13353264   0   34986930  A  C  \n",
       "\n",
       "[6478468 rows x 12 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge stat_sum with bim_df\n",
    "meta_bim = pd.merge(stat_sum, bim_df, on=['MarkerName'], how='inner')\n",
    "meta_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of unique rsid in duplicated meta_bim:  0\n",
      "number of unique rsid in meta_bim:  6478468\n",
      "number of unique MarkerName in meta_bim:  6478463\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3011341</th>\n",
       "      <td>chr22:19503090</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0449</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.48050</td>\n",
       "      <td>22</td>\n",
       "      <td>rs370330113</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3011342</th>\n",
       "      <td>chr22:19503090</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0449</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.48050</td>\n",
       "      <td>22</td>\n",
       "      <td>rs138605059</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3214210</th>\n",
       "      <td>chr22:19172321</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.0183</td>\n",
       "      <td>0.52360</td>\n",
       "      <td>22</td>\n",
       "      <td>rs199687134</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3214211</th>\n",
       "      <td>chr22:19172321</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.0183</td>\n",
       "      <td>0.52360</td>\n",
       "      <td>22</td>\n",
       "      <td>rs1780641</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279286</th>\n",
       "      <td>chr22:19497230</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.07624</td>\n",
       "      <td>22</td>\n",
       "      <td>rs2521701</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279287</th>\n",
       "      <td>chr22:19497230</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.07624</td>\n",
       "      <td>22</td>\n",
       "      <td>rs117551893</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3950528</th>\n",
       "      <td>chr22:19390804</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0161</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>0.32640</td>\n",
       "      <td>22</td>\n",
       "      <td>rs141585725</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3950529</th>\n",
       "      <td>chr22:19390804</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0161</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>0.32640</td>\n",
       "      <td>22</td>\n",
       "      <td>rs11089274</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5019123</th>\n",
       "      <td>chr22:19643104</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0006</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.96150</td>\n",
       "      <td>22</td>\n",
       "      <td>rs145711952</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5019124</th>\n",
       "      <td>chr22:19643104</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0006</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.96150</td>\n",
       "      <td>22</td>\n",
       "      <td>rs9618649</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             MarkerName Allele1 Allele2  Effect  StdErr  P-value  chr  \\\n",
       "3011341  chr22:19503090       c       t -0.0449  0.0637  0.48050   22   \n",
       "3011342  chr22:19503090       c       t -0.0449  0.0637  0.48050   22   \n",
       "3214210  chr22:19172321       t       g  0.0117  0.0183  0.52360   22   \n",
       "3214211  chr22:19172321       t       g  0.0117  0.0183  0.52360   22   \n",
       "3279286  chr22:19497230       c       t -0.0466  0.0263  0.07624   22   \n",
       "3279287  chr22:19497230       c       t -0.0466  0.0263  0.07624   22   \n",
       "3950528  chr22:19390804       c       t -0.0161  0.0164  0.32640   22   \n",
       "3950529  chr22:19390804       c       t -0.0161  0.0164  0.32640   22   \n",
       "5019123  chr22:19643104       c       t -0.0006  0.0134  0.96150   22   \n",
       "5019124  chr22:19643104       c       t -0.0006  0.0134  0.96150   22   \n",
       "\n",
       "                rsid  kb       pos a1 a2  \n",
       "3011341  rs370330113   0  19503090  C  T  \n",
       "3011342  rs138605059   0  19503090  C  T  \n",
       "3214210  rs199687134   0  19172321  C  T  \n",
       "3214211    rs1780641   0  19172321  T  G  \n",
       "3279286    rs2521701   0  19497230  T  C  \n",
       "3279287  rs117551893   0  19497230  C  T  \n",
       "3950528  rs141585725   0  19390804  T  C  \n",
       "3950529   rs11089274   0  19390804  C  T  \n",
       "5019123  rs145711952   0  19643104  A  G  \n",
       "5019124    rs9618649   0  19643104  C  T  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check duplicated rsid\n",
    "meta_bim_dup = meta_bim[meta_bim['rsid'].duplicated(keep=False)]\n",
    "print('number of unique rsid in duplicated meta_bim: ', meta_bim_dup['rsid'].nunique())\n",
    "#check number of unique rsid in case\n",
    "print('number of unique rsid in meta_bim: ', meta_bim.rsid.nunique()) \n",
    "#check number of unique MarkerName in case\n",
    "print('number of unique MarkerName in meta_bim: ', meta_bim.MarkerName.nunique())\n",
    "\n",
    "#print gwas summery stat with SNP that has duplicated MarkerName\n",
    "snp_stat_dup = meta_bim[meta_bim['MarkerName'].isin(dup_marker)]\n",
    "snp_stat_dup\n",
    "#7 duplicated (MarkerName) position with different rsid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numbers of rows in meta data: 6478468\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>60320992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>135908647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.1203</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>120407145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.8221</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>29372878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.2141</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>164053059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2  Effect  StdErr  P-value  chr          SNP        pos\n",
       "0    chr1:60320992  g  a -0.0185  0.0210   0.3801    1  rs116406626   60320992\n",
       "1   chr8:135908647  a  g -0.0054  0.0123   0.6598    8   rs11992603  135908647\n",
       "2  chr10:120407145  c  t  0.0278  0.0179   0.1203   10    rs4336949  120407145\n",
       "3   chr11:29372878  t  c  0.0049  0.0218   0.8221   11   rs72638174   29372878\n",
       "4   chr3:164053059  a  g  0.0607  0.0488   0.2141    3  rs115079612  164053059"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop rows in meta if rsid is NaN \n",
    "#meta_df = snp_stat[snp_stat['SNP'].notna()]\n",
    "\n",
    "sumstat_df = meta_bim.copy()\n",
    "#change Allele1 and Allele2 to A1 and A2\n",
    "sumstat_df = sumstat_df.rename(columns={'Allele1':'A1', 'Allele2':'A2','rsid':'SNP'})\n",
    "sumstat_df = sumstat_df.drop(['a1', 'a2', 'kb'], axis=1)\n",
    "\n",
    "#reset index\n",
    "#meta_df = meta_df.reset_index(drop=True)\n",
    "\n",
    "#check numbers of rows in meta_df\n",
    "print(\"numbers of rows in meta data:\", sumstat_df.shape[0])\n",
    "\n",
    "sumstat_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>60320992</td>\n",
       "      <td>-0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>135908647</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>120407145</td>\n",
       "      <td>1.553073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.822100</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>29372878</td>\n",
       "      <td>0.224771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.214100</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>164053059</td>\n",
       "      <td>1.243852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478463</th>\n",
       "      <td>chr16:84525085</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.211700</td>\n",
       "      <td>16</td>\n",
       "      <td>rs2927754</td>\n",
       "      <td>84525085</td>\n",
       "      <td>-1.248387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478464</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>11</td>\n",
       "      <td>rs148423893</td>\n",
       "      <td>47906648</td>\n",
       "      <td>0.299517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478465</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>3</td>\n",
       "      <td>rs62263573</td>\n",
       "      <td>47855594</td>\n",
       "      <td>-3.314286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478466</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "      <td>3</td>\n",
       "      <td>rs149847890</td>\n",
       "      <td>194372962</td>\n",
       "      <td>0.536000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478467</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "      <td>19</td>\n",
       "      <td>rs13353264</td>\n",
       "      <td>34986930</td>\n",
       "      <td>1.357143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6478468 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName A1 A2  Effect  StdErr   P-value  chr          SNP  \\\n",
       "0          chr1:60320992  g  a -0.0185  0.0210  0.380100    1  rs116406626   \n",
       "1         chr8:135908647  a  g -0.0054  0.0123  0.659800    8   rs11992603   \n",
       "2        chr10:120407145  c  t  0.0278  0.0179  0.120300   10    rs4336949   \n",
       "3         chr11:29372878  t  c  0.0049  0.0218  0.822100   11   rs72638174   \n",
       "4         chr3:164053059  a  g  0.0607  0.0488  0.214100    3  rs115079612   \n",
       "...                  ... .. ..     ...     ...       ...  ...          ...   \n",
       "6478463   chr16:84525085  t  c -0.0387  0.0310  0.211700   16    rs2927754   \n",
       "6478464   chr11:47906648  c  t  0.0062  0.0207  0.765700   11  rs148423893   \n",
       "6478465    chr3:47855594  c  t -0.0348  0.0105  0.000944    3   rs62263573   \n",
       "6478466   chr3:194372962  a  g  0.0268  0.0500  0.591900    3  rs149847890   \n",
       "6478467   chr19:34986930  a  c  0.0209  0.0154  0.174900   19   rs13353264   \n",
       "\n",
       "               pos         Z  \n",
       "0         60320992 -0.880952  \n",
       "1        135908647 -0.439024  \n",
       "2        120407145  1.553073  \n",
       "3         29372878  0.224771  \n",
       "4        164053059  1.243852  \n",
       "...            ...       ...  \n",
       "6478463   84525085 -1.248387  \n",
       "6478464   47906648  0.299517  \n",
       "6478465   47855594 -3.314286  \n",
       "6478466  194372962  0.536000  \n",
       "6478467   34986930  1.357143  \n",
       "\n",
       "[6478468 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculate z score with effect size and standard error (z = (score-mean)/se)\n",
    "\n",
    "def z_score(a, b):\n",
    "    if b == 0:\n",
    "        return 0.0\n",
    "    return float(a)/b\n",
    "\n",
    "sumstat_df['Z'] = np.vectorize(z_score)(sumstat_df['Effect'], sumstat_df['StdErr'])\n",
    "\n",
    "sumstat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of meta data: (6478468, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs116406626</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs11992603</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4336949</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>1.553073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs72638174</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.224771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115079612</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>1.243852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP A1 A2         Z\n",
       "0  rs116406626  g  a -0.880952\n",
       "1   rs11992603  a  g -0.439024\n",
       "2    rs4336949  c  t  1.553073\n",
       "3   rs72638174  t  c  0.224771\n",
       "4  rs115079612  a  g  1.243852"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#function to move columns\n",
    "def movecol(df, cols_to_move=[], ref_col='', place='After'):\n",
    "    \n",
    "    cols = df.columns.tolist()\n",
    "    if place == 'After':\n",
    "        seg1 = cols[:list(cols).index(ref_col) + 1]\n",
    "        seg2 = cols_to_move\n",
    "    if place == 'Before':\n",
    "        seg1 = cols[:list(cols).index(ref_col)]\n",
    "        seg2 = cols_to_move + [ref_col]\n",
    "    \n",
    "    seg1 = [i for i in seg1 if i not in seg2]\n",
    "    seg3 = [i for i in cols if i not in seg1 + seg2]\n",
    "    \n",
    "    return(df[seg1 + seg2 + seg3])\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "sumstat = movecol(sumstat_df, cols_to_move=['SNP'], ref_col='MarkerName', place='After')\n",
    "\n",
    "#drop MarkerName, Effect, StdErr, P-value colunms \n",
    "sumstat = sumstat.drop(['MarkerName', 'Effect', 'StdErr', 'P-value', 'chr', 'pos'], axis=1)\n",
    "\n",
    "sumstat = sumstat.reset_index(drop=True)\n",
    "print(\"shape of meta data:\", sumstat.shape)\n",
    "sumstat.head()\n",
    "\n",
    "##sumstat is ready to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              MarkerName A1 A2  Effect  StdErr   P-value  chr          SNP  \\\n",
      "0          chr1:60320992  g  a -0.0185  0.0210  0.380100    1  rs116406626   \n",
      "1         chr8:135908647  a  g -0.0054  0.0123  0.659800    8   rs11992603   \n",
      "2        chr10:120407145  c  t  0.0278  0.0179  0.120300   10    rs4336949   \n",
      "3         chr11:29372878  t  c  0.0049  0.0218  0.822100   11   rs72638174   \n",
      "4         chr3:164053059  a  g  0.0607  0.0488  0.214100    3  rs115079612   \n",
      "...                  ... .. ..     ...     ...       ...  ...          ...   \n",
      "6478463   chr16:84525085  t  c -0.0387  0.0310  0.211700   16    rs2927754   \n",
      "6478464   chr11:47906648  c  t  0.0062  0.0207  0.765700   11  rs148423893   \n",
      "6478465    chr3:47855594  c  t -0.0348  0.0105  0.000944    3   rs62263573   \n",
      "6478466   chr3:194372962  a  g  0.0268  0.0500  0.591900    3  rs149847890   \n",
      "6478467   chr19:34986930  a  c  0.0209  0.0154  0.174900   19   rs13353264   \n",
      "\n",
      "               pos         Z      chisq  \n",
      "0         60320992 -0.880952   0.776077  \n",
      "1        135908647 -0.439024   0.192742  \n",
      "2        120407145  1.553073   2.412035  \n",
      "3         29372878  0.224771   0.050522  \n",
      "4        164053059  1.243852   1.547169  \n",
      "...            ...       ...        ...  \n",
      "6478463   84525085 -1.248387   1.558470  \n",
      "6478464   47906648  0.299517   0.089710  \n",
      "6478465   47855594 -3.314286  10.984490  \n",
      "6478466  194372962  0.536000   0.287296  \n",
      "6478467   34986930  1.357143   1.841837  \n",
      "\n",
      "[6478468 rows x 11 columns]\n",
      "lambda using chisq:  1.171208791208791\n",
      "lambda using p-value:  1.0206140350877193\n"
     ]
    }
   ],
   "source": [
    "#calculate chisq\n",
    "stat_sum_chi = sumstat_df.copy()\n",
    "stat_sum_chi['chisq'] = stat_sum_chi['Z']**2\n",
    "print(stat_sum_chi)\n",
    "\n",
    "#calculate lambda (median(z^2) / 0.455)\n",
    "lambda_stat_sum = ((stat_sum_chi['chisq'].median())/0.455)\n",
    "print(\"lambda using chisq: \", lambda_stat_sum)\n",
    "\n",
    "#calculate lambda (median(z^2) / 0.455)\n",
    "lambda_statsum = ((stat_sum_chi['P-value'].median())/0.456)\n",
    "print(\"lambda using p-value: \", lambda_statsum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save meta_df as txt -- run only once for download\n",
    "sumstat.to_csv(r'./data_folder/meta.txt', sep='\\t' ,index=False)\n",
    "#gwas sumary stat is ready to use"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run this after post processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of significant GWAS data: (5891, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>31054607</td>\n",
       "      <td>-8.393939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>121534732</td>\n",
       "      <td>6.178947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>5</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>102446462</td>\n",
       "      <td>-5.854369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>31023273</td>\n",
       "      <td>-8.520000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>4</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>951040</td>\n",
       "      <td>-11.072165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2  Effect  StdErr       P-value  chr         SNP  \\\n",
       "0   chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   16   rs1108431   \n",
       "1  chr10:121534732  t  c  0.0587  0.0095  5.506000e-10   10    rs196202   \n",
       "2   chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    5    rs257320   \n",
       "3   chr16:31023273  a  g -0.0852  0.0100  1.338000e-17   16  rs12447930   \n",
       "4      chr4:951040  g  t -0.1074  0.0097  1.054000e-28    4   rs3733344   \n",
       "\n",
       "         pos          Z  \n",
       "0   31054607  -8.393939  \n",
       "1  121534732   6.178947  \n",
       "2  102446462  -5.854369  \n",
       "3   31023273  -8.520000  \n",
       "4     951040 -11.072165  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##grab significant GWAS hits (p <= 5e-8)\n",
    "sig_meta = sumstat_df[sumstat_df['P-value']< 5e-8]\n",
    "#reset index\n",
    "sig_meta = sig_meta.reset_index(drop=True)\n",
    "print(\"shape of significant GWAS data:\", sig_meta.shape)\n",
    "\n",
    "sig_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_meta.pos.dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [MarkerName, A1, A2, Effect, StdErr, P-value, SNP, Z]\n",
       "Index: []"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if double marker are in sig_meta\n",
    "sig_meta_dup = sig_meta[sig_meta['MarkerName'].isin(double_marker)]\n",
    "sig_meta_dup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save sig_meta -- run only once\n",
    "sig_meta.to_csv(r'./data_folder/gwas_pos.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>879584</td>\n",
       "      <td>894689</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>895967</td>\n",
       "      <td>901095</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>901877</td>\n",
       "      <td>911245</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18236</th>\n",
       "      <td>21</td>\n",
       "      <td>47720095</td>\n",
       "      <td>47743789</td>\n",
       "      <td>ENSG00000160298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18237</th>\n",
       "      <td>21</td>\n",
       "      <td>47744036</td>\n",
       "      <td>47865682</td>\n",
       "      <td>ENSG00000160299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18238</th>\n",
       "      <td>21</td>\n",
       "      <td>47878812</td>\n",
       "      <td>47989926</td>\n",
       "      <td>ENSG00000160305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18239</th>\n",
       "      <td>21</td>\n",
       "      <td>48018875</td>\n",
       "      <td>48025121</td>\n",
       "      <td>ENSG00000160307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18240</th>\n",
       "      <td>21</td>\n",
       "      <td>48055079</td>\n",
       "      <td>48085036</td>\n",
       "      <td>ENSG00000160310</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18241 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       X.Chr     start       end               ID\n",
       "0          1     69091     70008  ENSG00000186092\n",
       "1          1    860260    879955  ENSG00000187634\n",
       "2          1    879584    894689  ENSG00000188976\n",
       "3          1    895967    901095  ENSG00000187961\n",
       "4          1    901877    911245  ENSG00000187583\n",
       "...      ...       ...       ...              ...\n",
       "18236     21  47720095  47743789  ENSG00000160298\n",
       "18237     21  47744036  47865682  ENSG00000160299\n",
       "18238     21  47878812  47989926  ENSG00000160305\n",
       "18239     21  48018875  48025121  ENSG00000160307\n",
       "18240     21  48055079  48085036  ENSG00000160310\n",
       "\n",
       "[18241 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bring in coordinate data\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "      <th>pos_less</th>\n",
       "      <th>pos_more</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>31054607</td>\n",
       "      <td>-8.393939</td>\n",
       "      <td>30554607</td>\n",
       "      <td>31554607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>121534732</td>\n",
       "      <td>6.178947</td>\n",
       "      <td>121034732</td>\n",
       "      <td>122034732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>5</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>102446462</td>\n",
       "      <td>-5.854369</td>\n",
       "      <td>101946462</td>\n",
       "      <td>102946462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>31023273</td>\n",
       "      <td>-8.520000</td>\n",
       "      <td>30523273</td>\n",
       "      <td>31523273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>4</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>951040</td>\n",
       "      <td>-11.072165</td>\n",
       "      <td>451040</td>\n",
       "      <td>1451040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5886</th>\n",
       "      <td>chr3:151114889</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>-0.0618</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>2.501000e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>rs17204501</td>\n",
       "      <td>151114889</td>\n",
       "      <td>-6.306122</td>\n",
       "      <td>150614889</td>\n",
       "      <td>151614889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5887</th>\n",
       "      <td>chr6:32587090</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0833</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>2.978000e-11</td>\n",
       "      <td>6</td>\n",
       "      <td>rs28383312</td>\n",
       "      <td>32587090</td>\n",
       "      <td>-6.664000</td>\n",
       "      <td>32087090</td>\n",
       "      <td>33087090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5888</th>\n",
       "      <td>chr14:56192450</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.0557</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>2.820000e-09</td>\n",
       "      <td>14</td>\n",
       "      <td>rs7157327</td>\n",
       "      <td>56192450</td>\n",
       "      <td>-5.925532</td>\n",
       "      <td>55692450</td>\n",
       "      <td>56692450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5889</th>\n",
       "      <td>chr8:16735866</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0850</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>4.934000e-12</td>\n",
       "      <td>8</td>\n",
       "      <td>rs35847654</td>\n",
       "      <td>16735866</td>\n",
       "      <td>-6.910569</td>\n",
       "      <td>16235866</td>\n",
       "      <td>17235866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5890</th>\n",
       "      <td>chr1:171697497</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0665</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>1.949000e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>rs55670848</td>\n",
       "      <td>171697497</td>\n",
       "      <td>-5.635593</td>\n",
       "      <td>171197497</td>\n",
       "      <td>172197497</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5891 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           MarkerName A1 A2  Effect  StdErr       P-value  chr         SNP  \\\n",
       "0      chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   16   rs1108431   \n",
       "1     chr10:121534732  t  c  0.0587  0.0095  5.506000e-10   10    rs196202   \n",
       "2      chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    5    rs257320   \n",
       "3      chr16:31023273  a  g -0.0852  0.0100  1.338000e-17   16  rs12447930   \n",
       "4         chr4:951040  g  t -0.1074  0.0097  1.054000e-28    4   rs3733344   \n",
       "...               ... .. ..     ...     ...           ...  ...         ...   \n",
       "5886   chr3:151114889  g  a -0.0618  0.0098  2.501000e-10    3  rs17204501   \n",
       "5887    chr6:32587090  t  c -0.0833  0.0125  2.978000e-11    6  rs28383312   \n",
       "5888   chr14:56192450  c  t -0.0557  0.0094  2.820000e-09   14   rs7157327   \n",
       "5889    chr8:16735866  t  c -0.0850  0.0123  4.934000e-12    8  rs35847654   \n",
       "5890   chr1:171697497  t  c -0.0665  0.0118  1.949000e-08    1  rs55670848   \n",
       "\n",
       "            pos          Z   pos_less   pos_more  \n",
       "0      31054607  -8.393939   30554607   31554607  \n",
       "1     121534732   6.178947  121034732  122034732  \n",
       "2     102446462  -5.854369  101946462  102946462  \n",
       "3      31023273  -8.520000   30523273   31523273  \n",
       "4        951040 -11.072165     451040    1451040  \n",
       "...         ...        ...        ...        ...  \n",
       "5886  151114889  -6.306122  150614889  151614889  \n",
       "5887   32587090  -6.664000   32087090   33087090  \n",
       "5888   56192450  -5.925532   55692450   56692450  \n",
       "5889   16735866  -6.910569   16235866   17235866  \n",
       "5890  171697497  -5.635593  171197497  172197497  \n",
       "\n",
       "[5891 rows x 12 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create pos_less and pos_more which are position plus minus 1 megabase\n",
    "sig_meta['pos_less']= sig_meta['pos'] - 500000\n",
    "sig_meta['pos_more']= sig_meta['pos'] + 500000\n",
    "sig_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the 90 parkinson's GWAS hits from Mike's paper (some are in the supplement) ##https://www.thelancet.com/journals/laneur/article/PIIS1474-4422(19)30320-5/fulltext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read 90 parkinson's GWAS hits\n",
    "pd_hits = pd.read_csv(r\"./data_folder/PD_GRS_dbSNPs.tab\", sep='\\t', lineterminator='\\r', header=None,  names=('rsid', 'author', 'condition', 'name') )\n",
    "pd_hits.drop(pd_hits.iloc[:,1:4], axis=1, inplace=True)\n",
    "pd_hits['rsid'] = pd_hits['rsid'].str.replace('\\n', '')\n",
    "pd_hits['rsid'].replace('', np.nan, inplace=True)\n",
    "pd_hits.dropna(subset=['rsid'], inplace=True)\n",
    "pd_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge rsid with pos from meta data\n",
    "\n",
    "pd_hits_pos_1 = pd.merge(bim_df, pd_hits, on=['rsid'], how='inner')\n",
    "pd_hits_pos_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the rsid without pos\n",
    "mis_rsid = pd_hits[~pd_hits['rsid'].isin(pd_hits_pos_1['rsid'])]\n",
    "mis_rsid = mis_rsid.reset_index(drop=True)\n",
    "mis_rsid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge rsid with pos from meta data\n",
    "\n",
    "pd_hits_pos_2 = pd_hits.merge(bim_df, on=['rsid'], how='left') \n",
    "pd_hits_pos_2 = pd_hits_pos_2.drop(['kb'], axis=1)\n",
    "pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input data for rs114138760\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'chr'] = '1'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'pos'] = '154898185'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'MarkerName'] = 'chr1:154898185'\n",
    "\n",
    "#input data for rs6658353\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'chr'] = '1'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'pos'] = '161469054'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'MarkerName'] = 'chr1:161469054'\n",
    "\n",
    "#input data for rs76116224\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'pos'] = '18147848'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'MarkerName'] = 'chr2:18147848'\n",
    "\n",
    "#input data for rs2042477\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'pos'] = '96000943'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'MarkerName'] = 'chr2:96000943'\n",
    "\n",
    "#input data for rs11683001\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'pos'] = '102396963'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'MarkerName'] = 'chr2:102396963'\n",
    "\n",
    "#input data for rs11707416\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'chr'] = '3'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'pos'] = '151108965'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'MarkerName'] = 'chr3:151108965'\n",
    "\n",
    "#input data for rs34025766\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'chr'] = '4'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'pos'] = '17968811'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'MarkerName'] = 'chr4:17968811'\n",
    "\n",
    "#input data for rs6825004\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'chr'] = '4'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'pos'] = '77110365'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'a1'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'MarkerName'] = 'chr4:77110365'\n",
    "\n",
    "#input data for rs26431\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'chr'] = '5'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'pos'] = '102365794'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'MarkerName'] = 'chr5:102365794'\n",
    "\n",
    "#input data for rs76949143\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'chr'] = '7'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'pos'] = '66009851'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'MarkerName'] = 'chr7:66009851'\n",
    "\n",
    "#input data for rs6476434\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'chr'] = '9'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'pos'] = '34046391'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'MarkerName'] = 'chr9:34046391'\n",
    "\n",
    "#input data for rs6497339\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'chr'] = '16'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'pos'] = '19277493'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'MarkerName'] = 'chr16:19277493'\n",
    "\n",
    "#input data for rs2904880\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'chr'] = '16'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'pos'] = '28944396'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'a1'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'MarkerName'] = 'chr16:28944396'\n",
    "\n",
    "#input data for rs666463\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'chr'] = '17'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'pos'] = '76425480'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'MarkerName'] = 'chr17:76425480'\n",
    "\n",
    "#convert column values to int\n",
    "col_case=pd_hits_pos_2.columns[1:3]\n",
    "pd_hits_pos_2.loc[:,col_case]=pd_hits_pos_2.loc[:,col_case].astype(np.int64)\n",
    "\n",
    "pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save pd_hits_pos_2 \n",
    "pd_hits_pos_2.to_csv(r'./data_folder/PD_GRS_dbSNPs_position.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create pos_less and pos_more which are position plus minus 1 megabase\n",
    "pd_hits_pos_2['pos_less']= pd_hits_pos_2['pos'] - 1000000\n",
    "pd_hits_pos_2['pos_more']= pd_hits_pos_2['pos'] + 1000000\n",
    "pd_hits_pos_2\n",
    "#read 90 parkinson's GWAS hits\n",
    "#pd_hits_pos_2 = pd.read_csv(r\"./data_folder/PD_GRS_dbSNPs_position.csv\")\n",
    "#pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>pos_less</th>\n",
       "      <th>pos_more</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs114138760</td>\n",
       "      <td>1</td>\n",
       "      <td>154898185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:154898185</td>\n",
       "      <td>153898185</td>\n",
       "      <td>155898185</td>\n",
       "      <td>ENSG00000163344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs35749011</td>\n",
       "      <td>1</td>\n",
       "      <td>155135036</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:155135036</td>\n",
       "      <td>154135036</td>\n",
       "      <td>156135036</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>1</td>\n",
       "      <td>155205634</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:155205634</td>\n",
       "      <td>154205634</td>\n",
       "      <td>156205634</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6658353</td>\n",
       "      <td>1</td>\n",
       "      <td>161469054</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:161469054</td>\n",
       "      <td>160469054</td>\n",
       "      <td>162469054</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11578699</td>\n",
       "      <td>1</td>\n",
       "      <td>171719769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:171719769</td>\n",
       "      <td>170719769</td>\n",
       "      <td>172719769</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs12456492</td>\n",
       "      <td>18</td>\n",
       "      <td>40673380</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr18:40673380</td>\n",
       "      <td>39673380</td>\n",
       "      <td>41673380</td>\n",
       "      <td>ENSG00000152214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs8087969</td>\n",
       "      <td>18</td>\n",
       "      <td>48683589</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr18:48683589</td>\n",
       "      <td>47683589</td>\n",
       "      <td>49683589</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs55818311</td>\n",
       "      <td>19</td>\n",
       "      <td>2341047</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr19:2341047</td>\n",
       "      <td>1341047</td>\n",
       "      <td>3341047</td>\n",
       "      <td>ENSG00000005206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs77351827</td>\n",
       "      <td>20</td>\n",
       "      <td>6006041</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr20:6006041</td>\n",
       "      <td>5006041</td>\n",
       "      <td>7006041</td>\n",
       "      <td>ENSG00000088766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs2248244</td>\n",
       "      <td>21</td>\n",
       "      <td>38852361</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21:38852361</td>\n",
       "      <td>37852361</td>\n",
       "      <td>39852361</td>\n",
       "      <td>ENSG00000157540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid  chr        pos a1 a2      MarkerName   pos_less   pos_more  \\\n",
       "0   rs114138760    1  154898185  C  G  chr1:154898185  153898185  155898185   \n",
       "1    rs35749011    1  155135036  A  G  chr1:155135036  154135036  156135036   \n",
       "2    rs76763715    1  155205634  C  T  chr1:155205634  154205634  156205634   \n",
       "3     rs6658353    1  161469054  C  G  chr1:161469054  160469054  162469054   \n",
       "4    rs11578699    1  171719769  T  C  chr1:171719769  170719769  172719769   \n",
       "..          ...  ...        ... .. ..             ...        ...        ...   \n",
       "85   rs12456492   18   40673380  G  A  chr18:40673380   39673380   41673380   \n",
       "86    rs8087969   18   48683589  G  T  chr18:48683589   47683589   49683589   \n",
       "87   rs55818311   19    2341047  C  T   chr19:2341047    1341047    3341047   \n",
       "88   rs77351827   20    6006041  T  C   chr20:6006041    5006041    7006041   \n",
       "89    rs2248244   21   38852361  A  G  chr21:38852361   37852361   39852361   \n",
       "\n",
       "                 ID  \n",
       "0   ENSG00000163344  \n",
       "1               NaN  \n",
       "2   ENSG00000177628  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "..              ...  \n",
       "85  ENSG00000152214  \n",
       "86              NaN  \n",
       "87  ENSG00000005206  \n",
       "88  ENSG00000088766  \n",
       "89  ENSG00000157540  \n",
       "\n",
       "[90 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "result_1\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start']<= row['pos_more'] <= coords['end'] or coords['end'] >= row['pos_less'] >= coords['start']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "result_2\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> row['pos_less'] <= coords['start'] <= row['pos_more'] or row['pos_less'] <= coords['end'] <= row['pos_more']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "result_3\n",
    "\n",
    "###row['pos_less'] <= coords['start'] <= row['pos_more'] or row['pos_less'] <= coords['end'] <= row['pos_more']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in reult_1:  25\n",
      "Non Null ID in reult_1:  ['ENSG00000163344', 'ENSG00000177628', 'ENSG00000117280', 'ENSG00000143772', 'ENSG00000116991', 'ENSG00000170745', 'ENSG00000115041', 'ENSG00000071054', 'ENSG00000152128', 'ENSG00000131374', 'ENSG00000068745', 'ENSG00000114030', 'ENSG00000144893', 'ENSG00000196542', 'ENSG00000078070', 'ENSG00000178950', 'ENSG00000127419', 'ENSG00000109743', 'ENSG00000178177', 'ENSG00000138760', 'ENSG00000189157', 'ENSG00000189157', 'ENSG00000109572', 'ENSG00000164181', 'ENSG00000145730', 'ENSG00000204614', 'ENSG00000136235', 'ENSG00000164733', 'ENSG00000147439', 'ENSG00000153310', 'ENSG00000107295', 'ENSG00000107295', 'ENSG00000137073', 'ENSG00000077943', 'ENSG00000107862', 'ENSG00000151929', 'ENSG00000198825', 'ENSG00000110315', 'ENSG00000150672', 'ENSG00000080854', 'ENSG00000188906', 'ENSG00000188906', 'ENSG00000130787', 'ENSG00000102547', 'ENSG00000151338', 'ENSG00000131979', 'ENSG00000198208', 'ENSG00000103528', 'ENSG00000177548', 'ENSG00000099381', 'ENSG00000167207', 'ENSG00000170175', 'ENSG00000141699', 'ENSG00000108312', 'ENSG00000161682', 'ENSG00000120088', 'ENSG00000120088', 'ENSG00000108379', 'ENSG00000136492', 'ENSG00000187775', 'ENSG00000141431', 'ENSG00000152214', 'ENSG00000005206', 'ENSG00000088766', 'ENSG00000157540']\n"
     ]
    }
   ],
   "source": [
    "print('Null ID in reult_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in reult_1: ',result_1[['ID']].T.stack().tolist())\n",
    "#result_1[['ID']].T.stack().values\n",
    "\n",
    "print('Null ID in reult_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in reult_2: ',result_2[['ID']].T.stack().tolist())\n",
    "#result_1[['ID']].T.stack().values\n",
    "\n",
    "print('Null ID in reult_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in reult_3: ',result_3[['ID']].T.stack().tolist())\n",
    "#result_1[['ID']].T.stack().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop all columns except ID\n",
    "gwas_genes_1 = result_1.copy()\n",
    "gwas_genes_1.drop(gwas_genes_1.iloc[:,0:8], axis=1, inplace=True)\n",
    "gwas_genes_1.dropna(inplace=True)\n",
    "gwas_genes_1 = gwas_genes_1.reset_index(drop=True)\n",
    "\n",
    "gwas_genes_2 = result_2.copy()\n",
    "gwas_genes_2.drop(gwas_genes_2.iloc[:,0:8], axis=1, inplace=True)\n",
    "gwas_genes_2.dropna(inplace=True)\n",
    "gwas_genes_2 = gwas_genes_2.reset_index(drop=True)\n",
    "\n",
    "gwas_genes_3 = result_3.copy()\n",
    "gwas_genes_3.drop(gwas_genes_3.iloc[:,0:8], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ID in result_1:  65\n",
      "number of ID in result_2:  40\n",
      "number of ID in result_3:  90\n"
     ]
    }
   ],
   "source": [
    "#check common IDs between result_1~3\n",
    "gene_1 = gwas_genes_1.ID.tolist()\n",
    "gene_2 = gwas_genes_2.ID.tolist()\n",
    "gene_3 = result_3.ID.tolist()\n",
    "\n",
    "print('number of ID in result_1: ', len(gene_1))\n",
    "print('number of ID in result_2: ', len(gene_2))\n",
    "print('number of ID in result_3: ', len(gene_3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ID in result_1 and result_2:  3\n",
      "number of ID in result_1 and result_3:  4\n",
      "number of ID in result_2 and result_3:  40\n"
     ]
    }
   ],
   "source": [
    "def comprehension(a, b):\n",
    "    return [x for x in a if x in b]\n",
    "\n",
    "list_1 = comprehension(gene_1,gene_2)\n",
    "list_2 = comprehension(gene_1,gene_3)\n",
    "list_3 = comprehension(gene_2,gene_3)\n",
    "\n",
    "print('number of ID in result_1 and result_2: ', len(list_1))\n",
    "print('number of ID in result_1 and result_3: ', len(list_2))\n",
    "print('number of ID in result_2 and result_3: ', len(list_3))\n",
    "#since all of gene_2 are in gene_3 we will combine gene_1 and gene_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(142, 1)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge gwas_genes_1 and gwas_genes_3\n",
    "gwas = pd.merge(gwas_genes_1, gwas_genes_3, on='ID', how='outer')\n",
    "gwas = gwas.drop_duplicates(['ID']).reset_index(drop=True)\n",
    "gwas.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save gwas\n",
    "gwas.to_csv(r'./data_folder/PD_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
