{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GWAS summery stat data preprocess\n",
    "1.need to convert Markername to rsID by combining Chr and basepair coordinate in _lifted bim file and and replace Markername by matching\n",
    "2.calculate z-score using P-value and effect size using munge_sumstat in ldsc github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MarkerName      Allele1 Allele2 Freq1   FreqSE  MinFreq MaxFreq Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal freqSpan\n",
      "chr11:88249377  t       c       0.9905  0.0026  0.9882  0.9971  0.0238  0.0594  0.6889  -+?+++--+???-+++?       20      13.753  11      0.247   0.00890000000000002\n",
      "chr1:60320992   a       g       0.9375  0.0049  0.9221  0.9473  0.0185  0.021   0.3801  +-?--+?++??-+---+       20.1    15.022  12      0.2402  0.0252\n",
      "chr8:135908647  a       g       0.2112  0.007   0.1703  0.2316  -0.0054 0.0123  0.6598  --?+++?--??-----+       27.2    16.492  12      0.1697  0.0613\n",
      "chr16:77148858  a       g       0.9963  5e-04   0.9961  0.9984  -0.0228 0.1356  0.8663  -??-+??+-???----?       0       4.331   8       0.8261  0.00229999999999997\n",
      "chr11:97895884  c       g       0.0587  0.0035  0.0361  0.0685  -0.0219 0.0217  0.3148  +-?--+?--??+---++       0.2     12.029  12      0.4433  0.0324\n",
      "chr2:80117945   a       g       0.9979  7e-04   0.9968  0.9988  -0.4366 0.4506  0.3326  ?????+?--???---+?       0       5.236   6       0.5139  0.002\n",
      "chr18:814730    c       g       0.8893  0.0075  0.8639  0.9086  -0.0022 0.0176  0.9025  --?---?+-??+++--+       0       4.652   12      0.9686  0.0447\n",
      "chr19:32978300  a       g       0.0043  9e-04   0.0034  0.0075  0.0398  0.0991  0.6879  +-?--??--???+--+?       0       4.429   9       0.881   0.0041\n",
      "chr2:166277422  a       g       0.0023  0.0022  0.0011  0.0076  -0.1177 0.308   0.7023  ???+---?+????+???       21.5    6.372   5       0.2717  0.0065\n",
      "\u001b[K:\u001b[K20407145 t       c       0.9112  0.0071  0.8824  0.9202  -0.0278 0.0179  \u001b[7m/data/CARD/PD/summary_stats/META5_all.gz\u001b[m\u001b[K\u0007"
     ]
    }
   ],
   "source": [
    "#read meta5 file -- this is our GWAS summery stat data\n",
    "! zless /data/CARD/PD/summary_stats/META5_all.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy meta 5 file into data folder - run only once for download\n",
    "! cp -R /data/CARD/PD/summary_stats/META5_all.gz /data/songy4/TWAS/data/\n",
    "\n",
    "#unzip META5_all.gz - run only once for download\n",
    "! gzip -d META5_all.gz\n",
    "\n",
    "#make sumstat.csv which contains MarkerName, Allele1, Allele2, Effect, StdErr, P-value\n",
    "! awk '{print $1, $2, $3, $8, $9, $10}' data_folder/META5_all > data_folder/sumstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n"
     ]
    }
   ],
   "source": [
    "! awk '{print $1, $2, $3, $8, $9, $10}' data_folder/META5_all > data_folder/sumstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import library\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set paths\n",
    "basedir = '/data/songy4/twas'\n",
    "datadir = f'{basedir}/data_folder'\n",
    "fusiondir = f'{basedir}/fusion_twas'\n",
    "geno_path = f'{datadir}/qc_genotypes_twas'\n",
    "gene_list_path = f'{datadir}/gene_list.txt'\n",
    "pheno_path = f'{datadir}/expression_matrix_final.txt'\n",
    "coord_path = f'{datadir}/twas_coordinate.txt'\n",
    "covar_path = f'{datadir}/covariates.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       MarkerName Allele1 Allele2  Effect  StdErr  P-value\n",
      "0  chr11:88249377       t       c  0.0238  0.0594   0.6889\n",
      "1   chr1:60320992       a       g  0.0185  0.0210   0.3801\n",
      "2  chr8:135908647       a       g -0.0054  0.0123   0.6598\n",
      "3  chr16:77148858       a       g -0.0228  0.1356   0.8663\n",
      "4  chr11:97895884       c       g -0.0219  0.0217   0.3148\n",
      "shape of meta data: (2721631, 6)\n",
      "Unique MarkerName in meta data: 2721631\n"
     ]
    }
   ],
   "source": [
    "#check sumstat.csv\n",
    "meta = pd.read_csv(r\"./data_folder/sumstat.txt\", sep=' ')\n",
    "print(meta.head())\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"shape of meta data:\", meta.shape)\n",
    "\n",
    "#check the unique number of MarkerName in meta\n",
    "print(\"Unique MarkerName in meta data:\", meta['MarkerName'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs145427775</td>\n",
       "      <td>0</td>\n",
       "      <td>10291</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs55998931</td>\n",
       "      <td>0</td>\n",
       "      <td>10492</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199896944</td>\n",
       "      <td>0</td>\n",
       "      <td>13504</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>0</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0</td>\n",
       "      <td>14948</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr         rsid  kb    pos a1 a2\n",
       "0    1  rs145427775   0  10291  T  C\n",
       "1    1   rs55998931   0  10492  T  C\n",
       "2    1  rs199896944   0  13504  A  G\n",
       "3    1  rs199856693   0  14933  A  G\n",
       "4    1  rs201855936   0  14948  A  G"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open bim file \n",
    "bim_df = pd.read_csv(r\"./data_folder/qc_genotypes_twas_hg19_lifted.bim\", sep='\\t', header=None,  names=('chr', 'rsid', 'kb', 'pos', 'a1', 'a2'))\n",
    "bim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of bim data: (23060234, 7)\n",
      "number of unique MarkerName in bim data: 23060188\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs145427775</td>\n",
       "      <td>0</td>\n",
       "      <td>10291</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs55998931</td>\n",
       "      <td>0</td>\n",
       "      <td>10492</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199896944</td>\n",
       "      <td>0</td>\n",
       "      <td>13504</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:13504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>0</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0</td>\n",
       "      <td>14948</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr         rsid  kb    pos a1 a2  MarkerName\n",
       "0    1  rs145427775   0  10291  T  C  chr1:10291\n",
       "1    1   rs55998931   0  10492  T  C  chr1:10492\n",
       "2    1  rs199896944   0  13504  A  G  chr1:13504\n",
       "3    1  rs199856693   0  14933  A  G  chr1:14933\n",
       "4    1  rs201855936   0  14948  A  G  chr1:14948"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine chr and pos columns\n",
    "bim_df['MarkerName'] = str('chr') + bim_df.chr.astype(str) + ':' + bim_df.pos.astype(str)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"shape of bim data:\", bim_df.shape)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"number of unique MarkerName in bim data:\", bim_df['MarkerName'].nunique())\n",
    "\n",
    "bim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of NaN in rsid in meta data: 591918\n",
      "number of rsid in meta data: 2129713\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "      <td>rs11020170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>rs116406626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>rs11992603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "      <td>rs189372368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.3148</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName Allele1 Allele2  Effect  StdErr  P-value          SNP\n",
       "0  chr11:88249377       t       c  0.0238  0.0594   0.6889   rs11020170\n",
       "1   chr1:60320992       a       g  0.0185  0.0210   0.3801  rs116406626\n",
       "2  chr8:135908647       a       g -0.0054  0.0123   0.6598   rs11992603\n",
       "3  chr16:77148858       a       g -0.0228  0.1356   0.8663  rs189372368\n",
       "4  chr11:97895884       c       g -0.0219  0.0217   0.3148          NaN"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#add rsid column in meta by matching MarkerName from bim_df \n",
    "rename_dict = bim_df.set_index('MarkerName').to_dict()['rsid']\n",
    "meta['SNP'] = meta['MarkerName'].map(rename_dict)\n",
    "\n",
    "#count how many NaN in rsid in meta\n",
    "print(\"number of NaN in rsid in meta data:\", meta['SNP'].isna().sum())\n",
    "print(\"number of rsid in meta data:\", meta['SNP'].count())\n",
    "\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numbers of rows in meta data: 2129713\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "      <td>rs11020170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>rs116406626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>rs11992603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "      <td>rs189372368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19:32978300</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0398</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>0.6879</td>\n",
       "      <td>rs62106378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName A1 A2  Effect  StdErr  P-value          SNP\n",
       "0  chr11:88249377  t  c  0.0238  0.0594   0.6889   rs11020170\n",
       "1   chr1:60320992  a  g  0.0185  0.0210   0.3801  rs116406626\n",
       "2  chr8:135908647  a  g -0.0054  0.0123   0.6598   rs11992603\n",
       "3  chr16:77148858  a  g -0.0228  0.1356   0.8663  rs189372368\n",
       "4  chr19:32978300  a  g  0.0398  0.0991   0.6879   rs62106378"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop rows in meta if rsid is NaN \n",
    "meta_df = meta[meta['SNP'].notna()]\n",
    "\n",
    "#change Allele1 and Allele2 to A1 and A2\n",
    "meta_df = meta_df.rename(columns={'Allele1':'A1', 'Allele2':'A2'})\n",
    "\n",
    "#reset index\n",
    "meta_df = meta_df.reset_index(drop=True)\n",
    "\n",
    "#check numbers of rows in meta_df\n",
    "print(\"numbers of rows in meta data:\", meta_df.shape[0])\n",
    "\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "      <td>rs11020170</td>\n",
       "      <td>0.400673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "      <td>rs189372368</td>\n",
       "      <td>-0.168142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19:32978300</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0398</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>0.6879</td>\n",
       "      <td>rs62106378</td>\n",
       "      <td>0.401615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName A1 A2  Effect  StdErr  P-value          SNP         Z\n",
       "0  chr11:88249377  t  c  0.0238  0.0594   0.6889   rs11020170  0.400673\n",
       "1   chr1:60320992  a  g  0.0185  0.0210   0.3801  rs116406626  0.880952\n",
       "2  chr8:135908647  a  g -0.0054  0.0123   0.6598   rs11992603 -0.439024\n",
       "3  chr16:77148858  a  g -0.0228  0.1356   0.8663  rs189372368 -0.168142\n",
       "4  chr19:32978300  a  g  0.0398  0.0991   0.6879   rs62106378  0.401615"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculate z score with effect size and standard error\n",
    "\n",
    "def z_score(a, b):\n",
    "    if b == 0:\n",
    "        return 0.0\n",
    "    return float(a)/b\n",
    "\n",
    "meta_df['Z'] = np.vectorize(z_score)(meta_df['Effect'], meta_df['StdErr'])\n",
    "\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of meta data: (2129713, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs11020170</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.400673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs116406626</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs11992603</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs189372368</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.168142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs62106378</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.401615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP A1 A2         Z\n",
       "0   rs11020170  t  c  0.400673\n",
       "1  rs116406626  a  g  0.880952\n",
       "2   rs11992603  a  g -0.439024\n",
       "3  rs189372368  a  g -0.168142\n",
       "4   rs62106378  a  g  0.401615"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#function to move columns\n",
    "def movecol(df, cols_to_move=[], ref_col='', place='After'):\n",
    "    \n",
    "    cols = df.columns.tolist()\n",
    "    if place == 'After':\n",
    "        seg1 = cols[:list(cols).index(ref_col) + 1]\n",
    "        seg2 = cols_to_move\n",
    "    if place == 'Before':\n",
    "        seg1 = cols[:list(cols).index(ref_col)]\n",
    "        seg2 = cols_to_move + [ref_col]\n",
    "    \n",
    "    seg1 = [i for i in seg1 if i not in seg2]\n",
    "    seg3 = [i for i in cols if i not in seg1 + seg2]\n",
    "    \n",
    "    return(df[seg1 + seg2 + seg3])\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "sumstat = movecol(meta_df, cols_to_move=['SNP'], ref_col='MarkerName', place='After')\n",
    "\n",
    "#drop MarkerName, Effect, StdErr, P-value colunms \n",
    "sumstat = sumstat.drop(['MarkerName', 'Effect', 'StdErr', 'P-value'], axis=1)\n",
    "\n",
    "sumstat = sumstat.reset_index(drop=True)\n",
    "print(\"shape of meta data:\", sumstat.shape)\n",
    "sumstat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save meta_df as txt -- run only once for download\n",
    "sumstat.to_csv(r'./data_folder/meta_1.txt', sep='\\t' ,index=False)\n",
    "#gwas sumary stat is ready to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of significant GWAS data: (1381, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>-8.393939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>6.178947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>-5.854369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>-8.520000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>11.072165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2  Effect  StdErr       P-value         SNP          Z\n",
       "0   chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   rs1108431  -8.393939\n",
       "1  chr10:121534732  t  c  0.0587  0.0095  5.506000e-10    rs196202   6.178947\n",
       "2   chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    rs257320  -5.854369\n",
       "3   chr16:31023273  a  g -0.0852  0.0100  1.338000e-17  rs12447930  -8.520000\n",
       "4      chr4:951040  t  g  0.1074  0.0097  1.054000e-28   rs3733344  11.072165"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##grab significant GWAS hits (p <= 5e-8)\n",
    "sig_meta = meta_df[meta_df['P-value']<= 5e-8]\n",
    "#reset index\n",
    "sig_meta = sig_meta.reset_index(drop=True)\n",
    "print(\"shape of significant GWAS data:\", sig_meta.shape)\n",
    "\n",
    "sig_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>-8.393939</td>\n",
       "      <td>31054607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>6.178947</td>\n",
       "      <td>121534732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>-5.854369</td>\n",
       "      <td>102446462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>-8.520000</td>\n",
       "      <td>31023273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>11.072165</td>\n",
       "      <td>951040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2  Effect  StdErr       P-value         SNP          Z  \\\n",
       "0   chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   rs1108431  -8.393939   \n",
       "1  chr10:121534732  t  c  0.0587  0.0095  5.506000e-10    rs196202   6.178947   \n",
       "2   chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    rs257320  -5.854369   \n",
       "3   chr16:31023273  a  g -0.0852  0.0100  1.338000e-17  rs12447930  -8.520000   \n",
       "4      chr4:951040  t  g  0.1074  0.0097  1.054000e-28   rs3733344  11.072165   \n",
       "\n",
       "         pos  \n",
       "0   31054607  \n",
       "1  121534732  \n",
       "2  102446462  \n",
       "3   31023273  \n",
       "4     951040  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#add rsid column in meta by matching MarkerName from bim_df \n",
    "rename_dict = bim_df.set_index('rsid').to_dict()['pos']\n",
    "sig_meta['pos'] = sig_meta['SNP'].map(rename_dict)\n",
    "sig_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save sig_meta -- run only once\n",
    "sig_meta.to_csv(r'./data_folder/gwas_pos.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>860260</td>\n",
       "      <td>879955</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>879584</td>\n",
       "      <td>894689</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>895967</td>\n",
       "      <td>901095</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>901877</td>\n",
       "      <td>911245</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   X.Chr   start     end               ID\n",
       "0      1   69091   70008  ENSG00000186092\n",
       "1      1  860260  879955  ENSG00000187634\n",
       "2      1  879584  894689  ENSG00000188976\n",
       "3      1  895967  901095  ENSG00000187961\n",
       "4      1  901877  911245  ENSG00000187583"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bring in coords\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get ENSG ID from coords if pos in sig_meta are within start and stop in coords\n",
    "sig_meta.ID = sig_meta.merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "      <th>pos</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>-8.393939</td>\n",
       "      <td>31054607</td>\n",
       "      <td>ENSG00000169851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>6.178947</td>\n",
       "      <td>121534732</td>\n",
       "      <td>ENSG00000074047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>-5.854369</td>\n",
       "      <td>102446462</td>\n",
       "      <td>ENSG00000118733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>-8.520000</td>\n",
       "      <td>31023273</td>\n",
       "      <td>ENSG00000162949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>11.072165</td>\n",
       "      <td>951040</td>\n",
       "      <td>ENSG00000172554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>chr12:40727145</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.2018</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>1.638000e-11</td>\n",
       "      <td>rs28365229</td>\n",
       "      <td>6.726667</td>\n",
       "      <td>40727145</td>\n",
       "      <td>ENSG00000084073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>chr1:205739396</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0862</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>3.779000e-09</td>\n",
       "      <td>rs823139</td>\n",
       "      <td>-5.904110</td>\n",
       "      <td>205739396</td>\n",
       "      <td>ENSG00000117280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378</th>\n",
       "      <td>chr4:90630901</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.1540</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>1.091000e-54</td>\n",
       "      <td>rs356174</td>\n",
       "      <td>-15.555556</td>\n",
       "      <td>90630901</td>\n",
       "      <td>ENSG00000058091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379</th>\n",
       "      <td>chr6:32401261</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0773</td>\n",
       "      <td>0.0102</td>\n",
       "      <td>2.825000e-14</td>\n",
       "      <td>rs9268609</td>\n",
       "      <td>-7.578431</td>\n",
       "      <td>32401261</td>\n",
       "      <td>ENSG00000184007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1380</th>\n",
       "      <td>chr7:23113503</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.1014</td>\n",
       "      <td>0.0101</td>\n",
       "      <td>1.421000e-23</td>\n",
       "      <td>rs6461685</td>\n",
       "      <td>-10.039604</td>\n",
       "      <td>23113503</td>\n",
       "      <td>ENSG00000133216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1381 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           MarkerName A1 A2  Effect  StdErr       P-value         SNP  \\\n",
       "0      chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   rs1108431   \n",
       "1     chr10:121534732  t  c  0.0587  0.0095  5.506000e-10    rs196202   \n",
       "2      chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    rs257320   \n",
       "3      chr16:31023273  a  g -0.0852  0.0100  1.338000e-17  rs12447930   \n",
       "4         chr4:951040  t  g  0.1074  0.0097  1.054000e-28   rs3733344   \n",
       "...               ... .. ..     ...     ...           ...         ...   \n",
       "1376   chr12:40727145  t  c  0.2018  0.0300  1.638000e-11  rs28365229   \n",
       "1377   chr1:205739396  t  g -0.0862  0.0146  3.779000e-09    rs823139   \n",
       "1378    chr4:90630901  t  g -0.1540  0.0099  1.091000e-54    rs356174   \n",
       "1379    chr6:32401261  a  g -0.0773  0.0102  2.825000e-14   rs9268609   \n",
       "1380    chr7:23113503  t  c -0.1014  0.0101  1.421000e-23   rs6461685   \n",
       "\n",
       "              Z        pos               ID  \n",
       "0     -8.393939   31054607  ENSG00000169851  \n",
       "1      6.178947  121534732  ENSG00000074047  \n",
       "2     -5.854369  102446462  ENSG00000118733  \n",
       "3     -8.520000   31023273  ENSG00000162949  \n",
       "4     11.072165     951040  ENSG00000172554  \n",
       "...         ...        ...              ...  \n",
       "1376   6.726667   40727145  ENSG00000084073  \n",
       "1377  -5.904110  205739396  ENSG00000117280  \n",
       "1378 -15.555556   90630901  ENSG00000058091  \n",
       "1379  -7.578431   32401261  ENSG00000184007  \n",
       "1380 -10.039604   23113503  ENSG00000133216  \n",
       "\n",
       "[1381 rows x 10 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(row['pos'] <= coords['end']) & (row['pos'] >= coords['start'])]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.nan\n",
    "\n",
    "    return row\n",
    "\n",
    "result = sig_meta.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check how many id is missing\n",
    "result.ID.isna().sum()\n",
    "##total 1370 genes are significant in GWAS summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save result -- run only once\n",
    "result.to_csv(r'./data_folder/gwas_id.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000169851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000074047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000118733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000162949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000172554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>ENSG00000084073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>ENSG00000117280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378</th>\n",
       "      <td>ENSG00000058091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379</th>\n",
       "      <td>ENSG00000184007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1380</th>\n",
       "      <td>ENSG00000133216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1370 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID\n",
       "0     ENSG00000169851\n",
       "1     ENSG00000074047\n",
       "2     ENSG00000118733\n",
       "3     ENSG00000162949\n",
       "4     ENSG00000172554\n",
       "...               ...\n",
       "1376  ENSG00000084073\n",
       "1377  ENSG00000117280\n",
       "1378  ENSG00000058091\n",
       "1379  ENSG00000184007\n",
       "1380  ENSG00000133216\n",
       "\n",
       "[1370 rows x 1 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop all columns except ID\n",
    "gwas_genes = result.copy()\n",
    "gwas_genes.drop(gwas_genes.iloc[:,0:9], axis=1, inplace=True)\n",
    "gwas_genes.dropna(inplace=True)\n",
    "\n",
    "gwas_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save result -- run only once\n",
    "gwas_genes.to_csv(r'./data_folder/gwas_genes.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "compare these 1370 significant genes in GWAS summary stats with TWAS significant genes after TWAS pipeline is done."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
